frontend/catalyst/jit.py

```diff
@@ -439,6 +439,58 @@ def sum_abstracted(arr):
 
         the ``sum_abstracted`` function would only compile once and its definition would be
         reused for subsequent function calls.
+
+    .. _qubit_invariant_compilation:
+    .. details::
+        :title: Qubit-invariant compilation
+
+        In general, the inputs and outputs to the function being qjitted must have static types
+        and shapes. Otherwise, recompilation would be triggered.
+
+        Out of all the :ref:`Catalyst-supported terminal measurements <measurements>`, there are
+        four that have a return shape that depend on the number of qubits. Namely, the return shape
+        of :func:`qml.probs() <pennylane.probs>` and :func:`qml.state() <pennylane.state>` is a 1D
+        array of size ``(2**num_qubits)``, return shape of :func:`qml.sample() <pennylane.sample>`
+        is a 2D array of size ``(shots, num_qubits)``, and return shape of
+        :func:`qml.counts() <pennylane.counts>` is two 1D arrays of size ``(2**num_qubits)``.
+        The general rule of recompilation mentioned above means that when changing the number of qubits
+        in a workflow that returns these 4 measurements, recompilation would necessarily occur.
+
+        However, despite the general recompilation rule, Catalyst presents the powerful feature of
+        qubit-invariant compilation: the same compiled QNode can be invoked with a
+        different number of qubits! This is especially helpful for workflows where you would
```

Comment:
[Optional, not sure if this is the best wording]

```suggestion
        However, Catalyst offers a powerful exception to this rule with *qubit-invariant compilation*: the same compiled QNode can be invoked with a
        different number of qubits! This is especially helpful for workflows where you would
```